presto-main/src/main/java/com/facebook/presto/operator/scalar/RowDistinctFromOperator.java

```diff
@@ -0,0 +1,108 @@
+import java.util.List;
+
+import static com.facebook.presto.metadata.Signature.comparableWithVariadicBound;
+import static com.facebook.presto.spi.StandardErrorCode.GENERIC_INTERNAL_ERROR;
+import static com.facebook.presto.spi.function.OperatorType.IS_DISTINCT_FROM;
+import static com.facebook.presto.spi.type.TypeSignature.parseTypeSignature;
+import static com.facebook.presto.spi.type.TypeUtils.readNativeValue;
+import static com.facebook.presto.util.Reflection.methodHandle;
+
+public class RowDistinctFromOperator
+        extends SqlOperator
+{
+    public static final RowDistinctFromOperator ROW_IS_DISTINCT_FROM = new RowDistinctFromOperator();
+    private static final MethodHandle METHOD_HANDLE = methodHandle(RowDistinctFromOperator.class, "isDistinctFrom", Type.class, List.class, Block.class, Block.class);
+
+    private RowDistinctFromOperator()
+    {
+        super(IS_DISTINCT_FROM,
+                ImmutableList.of(comparableWithVariadicBound("T", "row")),
+                ImmutableList.of(),
+                parseTypeSignature(StandardTypes.BOOLEAN),
+                ImmutableList.of(parseTypeSignature("T"), parseTypeSignature("T")));
+    }
+
+    @Override
+    public ScalarFunctionImplementation specialize(BoundVariables boundVariables, int arity, TypeManager typeManager, FunctionRegistry functionRegistry)
+    {
+        ImmutableList.Builder<MethodHandle> argumentMethods = ImmutableList.builder();
+        Type type = boundVariables.getTypeVariable("T");
+        for (Type parameterType : type.getTypeParameters()) {
+            Signature signature = functionRegistry.resolveOperator(IS_DISTINCT_FROM, ImmutableList.of(parameterType, parameterType));
+            argumentMethods.add(functionRegistry.getScalarFunctionImplementation(signature).getMethodHandle());
+        }
+        return new ScalarFunctionImplementation(
+                false,  /* nullable */
+                ImmutableList.of(true, true),
+                METHOD_HANDLE.bindTo(type).bindTo(argumentMethods.build()),
+                isDeterministic());
+    }
+
+    public static boolean isDistinctFrom(Type rowType, List<MethodHandle> argumentMethods, Block leftRow, Block rightRow)
+    {
+        boolean leftNull = leftRow == null;
+        boolean rightNull = rightRow == null;
+        if (leftNull != rightNull) {
+            return true;
+        }
+        if (leftNull && rightNull) {
+            return false;
+        }
+        List<Type> fieldTypes = rowType.getTypeParameters();
+        for (int i = 0; i < leftRow.getPositionCount(); i++) {
+            boolean leftArgNull = leftRow.isNull(i);
+            boolean rightArgNull = rightRow.isNull(i);
+            if (leftArgNull != rightArgNull) {
+                return true;
+            }
+            if (leftArgNull && rightArgNull) {
+                continue;
+            }
+            Type type = fieldTypes.get(i);
+            try {
+                if ((Boolean) argumentMethods.get(i).invoke(
```

Comment:
I walked RowType.equals, and it also contains loop on varying type, which is also effectively megamorphics for row, right? With RowDistinctFrom this for loop in different place: operator itself, instead of type. It should be equal to row equal performance we have now.

```
        for (int i = 0; i < leftRow.getPositionCount(); i++) {
            Type fieldType = fields.get(i).getType();
            if (!fieldType.equalTo(leftRow, i, rightRow, i)) {
                return false;
            }
        }
```